```
 exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::ExifData::~ExifData(void)" (??1ExifData@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: unsigned __int64 __cdecl Exiv2::DataBuf::size(void)const " (?size@DataBuf@Exiv2@@QEBA_KXZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::DataBuf::~DataBuf(void)" (??1DataBuf@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: virtual __cdecl Exiv2::Error::~Error(void)" (??1Error@Exiv2@@UEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: bool __cdecl Exiv2::DataBuf::empty(void)const " (?empty@DataBuf@Exiv2@@QEBA_NXZ) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::DataBuf::DataBuf(void)" (??0DataBuf@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: struct Exiv2::DataBuf & __cdecl Exiv2::DataBuf::operator=(struct Exiv2::DataBuf &&)" (??4DataBuf@Exiv2@@QEAAAEAU01@$$QEAU01@@Z) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: class std::_Vector_iterator<class std::_Vector_val<struct std::_Simple_types<unsigned char> > > __cdecl Exiv2::DataBuf::begin(void)" (?begin@DataBuf@Exiv2@@QEAA?AV?$_Vector_iterator@V?$_Vector_val@U?$_Simple_types@E@std@@@std@@@std@@XZ) already defined in libexiv2int.a(src_cr2header_int.cpp.obj)
```

That is...very interesting

--- a/.github/workflows/on_PR_meson.yaml
+++ b/.github/workflows/on_PR_meson.yaml
@@ -27,6 +27,7 @@ jobs:
         run: |
           env CXX=g++-${{matrix.cxx}} meson setup "${{github.workspace}}/build" --wrap-mode=${{matrix.deps}} -Ddefault_library=${{matrix.type}}
           meson compile -C "${{github.workspace}}/build" --verbose
+          meson test -C "${{github.workspace}}/build" --verbose
 
   VisualStudio:
     runs-on: windows-latest
@@ -47,6 +48,7 @@ jobs:
         run: |
           meson setup "${{github.workspace}}/build" --wrap-mode=${{matrix.deps}} -Ddefault_library=${{matrix.type}}
           meson compile -C "${{github.workspace}}/build" --verbose
+          meson test -C "${{github.workspace}}/build" --verbose
 
   MacOS:
     runs-on: macos-latest
@@ -66,3 +68,4 @@ jobs:
         run: |
           meson setup "${{github.workspace}}/build" --wrap-mode=${{matrix.deps}} -Ddefault_library=${{matrix.type}}
           meson compile -C "${{github.workspace}}/build" --verbose
+          meson test -C "${{github.workspace}}/build" --verbose

--- a/.github/workflows/on_PR_meson_clang.yaml
+++ b/.github/workflows/on_PR_meson_clang.yaml
@@ -27,3 +27,4 @@ jobs:
         run: |
           env CXX=clang++-${{matrix.cxx}} CXXFLAGS=-stdlib=libc++ meson setup "${{github.workspace}}/build" --wrap-mode=${{matrix.deps}} -Ddefault_library=${{matrix.type}}
           meson compile -C "${{github.workspace}}/build" --verbose
+          meson test -C "${{github.workspace}}/build" --verbose

--- a/meson.build
+++ b/meson.build
@@ -339,3 +339,5 @@ executable(
   dependencies: exiv2_dep,
   install: true,
 )
+
+subdir('unitTests')

--- a/meson.build
+++ b/meson.build
@@ -274,15 +274,28 @@ install_headers(
   subdir: 'exiv2',
 )
 
+exiv2int = static_library(
+  'exiv2int',
+  int_lib,
+  cpp_args: '-DEXIV2API=',
+  dependencies: deps,
+  include_directories: libinc,
+)
+
+depinc = include_directories('.', 'include')
+exiv2int_dep = declare_dependency(
+  include_directories: depinc,
+  link_with: exiv2int,
+)
+
 exiv2 = library(
   'exiv2',
   base_lib,
-  int_lib,
   xmp_lib,
   cpp_args: cargs,
   version: meson.project_version(),
   gnu_symbol_visibility: 'hidden',
-  dependencies: deps,
+  dependencies: [deps, exiv2int_dep],
   include_directories: libinc,
   install: true,
 )
@@ -292,7 +305,6 @@ if host_machine.system() == 'windows' and get_option('default_library') != 'stat
   dllapi = '-DEXIV2API=__declspec(dllimport)'
 endif
 
-depinc = include_directories('include')
 exiv2_dep = declare_dependency(
   compile_args: dllapi,
   dependencies: intl_dep,

--- a/meson_options.txt
+++ b/meson_options.txt
@@ -40,3 +40,7 @@ option('video', type : 'boolean',
 option('xmp', type : 'feature',
   description : 'Build support for XMP',
 )
+
+option('unitTests', type : 'feature',
+  description : 'Build and run unit tests',
+)

--- a/subprojects/gtest.wrap
+++ b/subprojects/gtest.wrap
@@ -1,0 +1,15 @@
+[wrap-file]
+directory = googletest-1.13.0
+source_url = https://github.com/google/googletest/archive/refs/tags/v1.13.0.tar.gz
+source_filename = gtest-1.13.0.tar.gz
+source_hash = ad7fdba11ea011c1d925b3289cf4af2c66a352e18d4c7264392fead75e919363
+patch_filename = gtest_1.13.0-1_patch.zip
+patch_url = https://wrapdb.mesonbuild.com/v2/gtest_1.13.0-1/get_patch
+patch_hash = 6d82a02c3a45071cea989983bf6becde801cbbfd29196ba30dada0215393b082
+wrapdb_version = 1.13.0-1
+
+[provide]
+gtest = gtest_dep
+gtest_main = gtest_main_dep
+gmock = gmock_dep
+gmock_main = gmock_main_dep

--- a/unitTests/meson.build
+++ b/unitTests/meson.build
@@ -1,0 +1,57 @@
+gtest_dep = dependency('gmock_main', required: get_option('unitTests'))
+if not gtest_dep.found()
+  subdir_done()
+endif
+
+test_sources = files(
+  'test_DateValue.cpp',
+  'test_Error.cpp',
+  'test_FileIo.cpp',
+  'test_ImageFactory.cpp',
+  'test_IptcKey.cpp',
+  'test_LangAltValueRead.cpp',
+  'test_Photoshop.cpp',
+  'test_TimeValue.cpp',
+  'test_XmpKey.cpp',
+  'test_basicio.cpp',
+  'test_bmpimage.cpp',
+  'test_cr2header_int.cpp',
+  'test_datasets.cpp',
+  'test_enforce.cpp',
+  'test_futils.cpp',
+  'test_helper_functions.cpp',
+  'test_image_int.cpp',
+  'test_jp2image.cpp',
+  'test_jp2image_int.cpp',
+  'test_pngimage.cpp',
+  'test_safe_op.cpp',
+  'test_slice.cpp',
+  'test_tiffheader.cpp',
+  'test_types.cpp',
+  'test_utils.cpp',
+)
+
+if get_option('video')
+  test_sources += files(
+    'test_asfvideo.cpp',
+    'test_matroskavideo.cpp',
+    'test_riffVideo.cpp',
+  )
+endif
+
+if host_machine.system() == 'windows' and get_option('default_library') != 'static'
+  test_sources += int_lib
+endif
+
+t_args = ['-UEXIV2API', '-DEXIV2API=', '-DTESTDATA_PATH="@0@"'.format('..' / 'test' / 'data')]
+
+privinc = include_directories('../src', '../include/exiv2')
+unit_tests = executable(
+  'unit_tests',
+  test_sources,
+  cpp_args: t_args,
+  dependencies: [deps, exiv2_dep, gtest_dep],
+  include_directories: privinc,
+)
+
+test('Unit Tests', unit_tests)

